### DISCUSSION TITLE
How should I inject services into a class thats being used in axaml?

### DISCUSSION THREAD

**Discussion by XDgl233** [AvaloniaUI/Avalonia#14337] [Q&A]
This problem comes when I try to use the view locator (implementing IDataTemplate interface) with dependency injection using constructor injection. I can't find a proper way to inject services into the data templates in the App class from code, as the DataTemplates property is read-only. I try to use the AvaloniaXamlLoader.Load(IServiceProvider sp, this) overload, but it seems the constructor with parameters is still skipped and calls the default constructor instead.
So how can I resolve services in such cases, and maybe in a more generic scope, how should I resolve service from DI that's being called inside the axaml builder? Thanks :)

**Comment by timunie** [comment]
You can use Design.IsDesignTime to handle that.

**Reply by XDgl233** [reply]
Can I set datatemplate of the App from codes? So far if I inject the services through constructor, and apply the view locator in the App.axaml like [the doc](https://docs.avaloniaui.net/docs/concepts/view-locator#using-the-view-locator), the app crashes and reminds me it cannot resolve that class ("Unable to find public constructor...").

**Reply by timunie** [reply]
Yes, you can implement IDataTemplate and handle all of it on your own. https://github.com/AvaloniaUI/Avalonia.Samples?tab=readme-ov-file#%EF%B8%8F-datatemplate-samples

**Comment by thevortexcloud** [comment]
If you are trying to resolve a VM at design time, you can make a markup extension for that: https://github.com/AvaloniaUI/Avalonia/discussions/13743#discussioncomment-7686499
